[Merged by Bors] - Implement standard eth2.0 API

paulhauner opened the pull request:
## Issue Addressed

- Resolves #1550
- Resolves #824
- Resolves #825
- Resolves #1131
- Resolves #1411
- Resolves #1256
- Resolve #1177

## Proposed Changes

- Includes the `ShufflingId` struct initially defined in #1492. That PR is now closed and the changes are included here, with significant bug fixes.
- Implement the https://github.com/ethereum/eth2.0-APIs in a new `http_api` crate using `warp`. This replaces the `rest_api` crate.
- Add a new `common/eth2` crate which provides a wrapper around `reqwest`, providing the HTTP client that is used by the validator client and for testing. This replaces the `common/remote_beacon_node` crate.
- Create a `http_metrics` crate which is a dedicated server for Prometheus metrics (they are no longer served on the same port as the REST API). We now have flags for `--metrics`, `--metrics-address`, etc.
- Allow the `subnet_id` to be an optional parameter for `VerifiedUnaggregatedAttestation::verify`. This means it does not need to be provided unnecessarily by the validator client.
- Move `fn map_attestation_committee` in `mod beacon_chain::attestation_verification` to a new `fn with_committee_cache` on the `BeaconChain` so the same cache can be used for obtaining validator duties.
- Add some other helpers to `BeaconChain` to assist with common API duties (e.g., `block_root_at_slot`, `head_beacon_block_root`).
- Change the `NaiveAggregationPool` so it can index attestations by `hash_tree_root(attestation.data)`. This is a requirement of the API.
- Add functions to `BeaconChainHarness` to allow it to create slashings and exits.
- Allow for `eth1::Eth1NetworkId` to go to/from a `String`.
- Add functions to the `OperationPool` to allow getting all objects in the pool.
- Add function to `BeaconState` to check if a committee cache is initialized.
- Fix bug where `seconds_per_eth1_block` was not transferring over from `YamlConfig` to `ChainSpec`.
- Add the `deposit_contract_address` to `YamlConfig` and `ChainSpec`. We needed to be able to return it in an API response.
- Change some uses of serde `serialize_with` and `deserialize_with` to a single use of `with` (code quality).
- Impl `Display` and `FromStr` for several BLS fields.
- Check for clock discrepancy when VC polls BN for sync state (with +/- 1 slot tolerance). This is not intended to be comprehensive, it was just easy to do.

## Additional Info

- See #1434 for a per-endpoint overview.
- Seeking clarity here: https://github.com/ethereum/eth2.0-APIs/issues/75

## TODO

- [x] Add docs for prom port to close #1256
- [x] Follow up on this #1177
- [x] ~~Follow up with #1424~~ Will fix in future PR.
- [x] Follow up with #1411
- [x] ~~Follow up with  #1260~~ Will fix in future PR.
- [x] Add quotes to all integers.
- [x] Remove `rest_types`
- [x] Address missing beacon block error. (#1629)
- [x] ~~Add tests for lighthouse/peers endpoints~~ Wontfix
- [x] ~~Follow up with validator status proposal~~ Tracked in #1434
- [x] Unify graffiti structs
- [x] ~~Start server when waiting for genesis?~~ Will fix in future PR.
- [x] TODO in http_api tests
- [x] Move lighthouse endpoints off /eth/v1
- [x] Update docs to link to standard

## Blocked On

- ~~Blocked on #1586~~

michaelsproul commented:
I've been using your `QuotedIntVisitor` in #1544 and noticed that it only deserializes quoted integers, and errors on unquoted ones. This breaks the YAML parsing in the EF tests, but it can be fixed by adding an implementation of `visit_u64`:

```rust
fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
where
    E: serde::de::Error,
{
    Ok(v)
}
```

For the slashing interchange format I actually need it to error on quoted ints, so I made a variant `only_quoted` module that preserves the current behaviour. I was thinking we could have flexible quoting on by default for most of the types in `consensus/types`, and then opt into stricter behaviour where it's desired? Does that work for you?

paulhauner commented on validator_client/src/validator_duty.rs:
```diff
@@ -0,0 +1,125 @@
+use eth2::{
```
Add comments to this file.

paulhauner commented on validator_client/src/validator_duty.rs:
```diff
@@ -0,0 +1,125 @@
+
+/// This struct is being used as a shim since we deprecated the `rest_api` in favour of `http_api`.
+///
+/// TODO: add an issue about this.
```
TODO: make issue

paulhauner reviewed (commented):
Initial self review :)

paulhauner commented on book/src/advanced_metrics.md:
```diff
@@ -0,0 +1,34 @@
+# Prometheus Metrics
+
+Lighthouse provides and extensive suite of metrics and monitoring in the
```
`and` -> `an`

paulhauner commented on book/src/api-bn.md:
```diff
@@ -0,0 +1,95 @@
+	provided).
+- `--http-port`: specify the listen port of the server.
+- `--http-address`: specify the listen address of the server.
+
```
Add CLI example

paulhauner commented on book/src/api-bn.md:
````diff
@@ -0,0 +1,95 @@
+You can quickly check that the HTTP endpoint is up using `curl`:
+
+```
+curl "localhost:5052/beacon/head"
````
Fix this example

guybrush commented:
Currently some values are numbers which should be strings.

For example in the response of `/eth/v1/beacon/states/{stateID}/validators` `.data[].validator.{effective_balance,activation_eligibilty_epoch,activation_epoch,exit_epoch,withdrawable_epoch}` are numbers.

michaelsproul commented on beacon_node/beacon_chain/src/shuffling_cache.rs:
```diff
@@ -36,11 +36,38 @@ impl ShufflingCache {
+            Some(self.current.clone())
+        } else if epoch == self.next.shuffling_epoch {
+            Some(self.next.clone())
+        } else if epoch > self.next.shuffling_epoch {
+            let mut shuffling_id = self.next.clone();
+            shuffling_id.shuffling_epoch = epoch;
+            Some(shuffling_id)
+        } else {
```
This doesn't seem safe to me... if we have the shuffling IDs for the current and next epochs, why can we assume we know the shuffling for future epochs?

michaelsproul replied:
I kinda convinced myself that this was OK during our call, but I think it's still not quite right.

Our goal here is to return a shuffling ID that accurately describes the shuffling at a future epoch where no new blocks have been produced since the given head block. The problem is, the shuffling ID for the next epoch (from the perspective of the head), only includes RANDAO information up to the last slot of the previous epoch. The shuffling for subsequent epochs (>`next_epoch`) is affected by blocks _after_ that point, namely those in the current epoch, including the head block itself. Diagram for clarity:

![](https://i.imgur.com/oLjRtNN.jpg)

I think the correct shuffling ID to return for an `epoch` after `next_epoch` would be:

```rust
ShufflingId {
    shuffling_epoch: epoch,
    shuffling_decision_block: head_block_root,
}
```

Which would require incorporating the `head_block_root` into the `BlockShufflingIds` struct.

paulhauner replied:
Yep, you're totally right. Thank you!

michaelsproul commented on consensus/types/src/shuffling_id.rs:
```diff
@@ -0,0 +1,64 @@
+        // Taking advantage of saturating subtraction on slot and epoch.
+        //
+        // This is the final slot of the penultimate epoch.
+        let shuffling_decision_slot =
+            (state.current_epoch() - 1).start_slot(E::slots_per_epoch()) - 1;
```
Shouldn't this subtract from the `shuffling_epoch` in question instead of the current epoch? I.e.

```rust
(shuffling_epoch - 1).start_slot(E::slots_per_epoch()) - 1
```

Or even clearer:

```rust
(shuffling_epoch - 2).end_slot(E::slots_per_epoch())
```

We could also replace the 2 by `1 + spec.min_seed_lookahead`, to be robust against potential spec updates (unlikely, but possible).

paulhauner replied:
You're right about the `shuffling_epoch`, thanks!

> `(shuffling_epoch - 2).end_slot(E::slots_per_epoch())`

Unfortunately this doesn't work for cases like `state.slot ==1`, where you'll end up with a shuffling decision slot of `31`.

michaelsproul replied:
Ah, yeah!

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -1992,6 +2020,128 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         Ok(())
     }
 
+    /// Runs the `map_fn` with the committee cache at `shuffling_epoch` if the head of the chain is
+    /// `head_block_root`.
```
There's something strange about this comment, the "if" implies it might only work for the canonical head. Maybe instead we could say something like:

```
Runs the `map_fn` with the committee cache for `shuffling_epoch` from the chain with head `head_block_root`.
```

michaelsproul reviewed (commented):
(submitting early to discuss review-so-far)

michaelsproul commented on consensus/types/src/shuffling_id.rs:
```diff
@@ -0,0 +1,64 @@
+        let shuffling_decision_slot =
+            (state.current_epoch() - 1).start_slot(E::slots_per_epoch()) - 1;
+
+        let shuffling_decision_block = if state.slot == shuffling_decision_slot {
```
We discussed changing this to a `<=` on the call, but I think that isn't necessary because this function takes a relative epoch, and it's impossible for the `shuffling_decision_slot` to be `> state.slot` because it is at most `(state.next_epoch() - 2).end_slot(..)`, i.e. `state.previous_epoch().end_slot(..)`.

guybrush commented:
Another field deviating from the spec is `/eth/v1/beacon/states/{stateID}/validators.data[].validator.status`.

For example in the response for `/eth/v1/beacon/states/160/validators` I get:
```
...
{ index: '2329',
  balance: '31000063399',
  status: { ActiveAwaitingSlashedExit: 9 },
  validator:
   { pubkey:
      '0x880c00684c898d7b336168ea26c127a6b3ea2791a4b9298e33eaaf6cac0d4734a8d8f78dea41a2087a4b6c7cf847b46f',
     withdrawal_credentials:
      '0x00241e0a3179a16bdd36bcea3c7658c9eb12a1a444b0d67382411a692a1085d2',
     effective_balance: 31000000000,
     slashed: true,
     activation_eligibility_epoch: 0,
     activation_epoch: 0,
     exit_epoch: 9,
     withdrawable_epoch: 8196 } }
...
```

paulhauner commented:
> Another field deviating from the spec is /eth/v1/beacon/states/{stateID}/validators.data[].validator.status.

Thanks @guybrush, that is known and tracked here: https://github.com/sigp/lighthouse/issues/1434. Ultimately I'd like to change the specification.

Also, love the Monkey Island reference :D

realbigsean reviewed (changes requested):
Reviewed as best I could, this thing's a monster! Left small comments.

michaelsproul commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -0,0 +1,1693 @@
+    // there's a `canonical` flag on the response, I assume it should also return non-canonical
+    // things. Returning non-canonical things is hard for us since we don't already have a
+    // mechanism for arbitrary forwards block iteration, we only support iterating forwards along
+    // the canonical chain.
```
We can iterate forwards to a single arbitrary head, but that functionality isn't exposed by the `BeaconChain` at the moment. It looks like this API might need forwards iteration forwards across _all_ branches/heads, in which case we're out of luck?

paulhauner replied:
I've added a note to #1434 so we can get to this as we finish aligning with the spec.

michaelsproul commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -0,0 +1,1693 @@
+    // Define the ultimate set of routes that will be provided to the server.
+    let routes = warp::get()
+        .and(
+            get_beacon_genesis
+                .or(get_beacon_state_root.boxed())
+                .or(get_beacon_state_fork.boxed())
+                .or(get_beacon_state_finality_checkpoints.boxed())
+                .or(get_beacon_state_validators.boxed())
+                .or(get_beacon_state_validators_id.boxed())
+                .or(get_beacon_state_committees.boxed())
+                .or(get_beacon_headers.boxed())
+                .or(get_beacon_headers_block_id.boxed())
+                .or(get_beacon_block.boxed())
+                .or(get_beacon_block_attestations.boxed())
+                .or(get_beacon_block_root.boxed())
+                .or(get_beacon_pool_attestations.boxed())
+                .or(get_beacon_pool_attester_slashings.boxed())
+                .or(get_beacon_pool_proposer_slashings.boxed())
+                .or(get_beacon_pool_voluntary_exits.boxed())
+                .or(get_config_fork_schedule.boxed())
+                .or(get_config_spec.boxed())
+                .or(get_config_deposit_contract.boxed())
+                .or(get_debug_beacon_states.boxed())
+                .or(get_debug_beacon_heads.boxed())
+                .or(get_node_identity.boxed())
+                .or(get_node_version.boxed())
+                .or(get_node_syncing.boxed())
+                .or(get_validator_duties_attester.boxed())
+                .or(get_validator_duties_proposer.boxed())
+                .or(get_validator_blocks.boxed())
+                .or(get_validator_attestation_data.boxed())
+                .or(get_validator_aggregate_attestation.boxed())
+                .or(get_lighthouse_health.boxed())
+                .or(get_lighthouse_syncing.boxed())
+                .or(get_lighthouse_peers.boxed())
+                .or(get_lighthouse_peers_connected.boxed())
+                .or(get_lighthouse_proto_array.boxed())
+                .or(get_lighthouse_validator_inclusion_global.boxed())
+                .or(get_lighthouse_validator_inclusion.boxed())
+                .boxed(),
```
:exploding_head:

michaelsproul reviewed (changes requested):
Ooft, that was a big PR to review haha :sweat_smile: 

I had a good time though. Everything's well-laid out, neat, and superbly implemented :100:

I have just a few minor edits suggested before I think we're ready to ship this! (see my first review above too)

michaelsproul commented on beacon_node/http_api/src/validator_inclusion.rs:
```diff
@@ -0,0 +1,90 @@
+    chain: &BeaconChain<T>,
+) -> Result<GlobalValidatorInclusionData, warp::Rejection> {
+    // This is the last slot of the given epoch (one prior to the first slot of the next epoch).
+    let target_slot = (epoch + 1).start_slot(T::EthSpec::slots_per_epoch()) - 1;
```
(dupe of Sean's)

guybrush commented:
Oh I see, will just use what your current implementation gives me and adapt as needed then!

Look behind you, a three-headed monkey!

paulhauner commented on beacon_node/beacon_chain/src/test_utils.rs:
```diff
@@ -129,7 +131,7 @@ impl<E: EthSpec> BeaconChainHarness<BlockingMigratorEphemeralHarnessType<E>> {
         let decorator = slog_term::PlainDecorator::new(slog_term::TestStdoutWriter);
         let drain = slog_term::FullFormat::new(decorator).build();
-        let debug_level = slog::LevelFilter::new(drain, slog::Level::Debug);
+        let debug_level = slog::LevelFilter::new(drain, slog::Level::Critical);
```
I found that it was getting quite noisy when you're *not* testing things inside the harness/chain.

This was almost always the case for my HTTP API tests and I found myself scrolling through 6 epochs of harness logs to find my actual failures.

My thoughts were to keep the harness unobtrusive by default, with op-in verbose logging.

michaelsproul replied:
Cool, yeah, I'm fine with that

paulhauner commented on testing/simulator/src/cli.rs:
```diff
@@ -34,7 +34,7 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
                         .long("speed_up_factor")
                         .takes_value(true)
-                        .default_value("4")
+                        .default_value("3")
```
Oh yeah, this was an interesting one.. The initial seconds-per-slot was 6,000ms and when divided by 4 becomes `1500ms`.

When we use the `/config/spec` endpoint, our "millliseconds per slot" value gets turned into "seconds per slot" and back again. Due to this the VC gets a "milliseconds per slot" value of 1000ms instead of 1500ms and throws an error saying the BN is on a bad spec version.

It didn't quite seem worth it to fix this whole thing, so I took the easy route.

michaelsproul replied:
Interesting! I'm happy with that resolution

paulhauner commented on beacon_node/http_api/src/beacon_proposer_cache.rs:
```diff
@@ -0,0 +1,168 @@
+        let epoch_boundary_slot = head_state
+            .current_epoch()
+            .start_slot(T::EthSpec::slots_per_epoch());
+        let epoch_boundary_root = if head_state.slot >= epoch_boundary_slot {
```
The whole code block is:

```rust
        let epoch_boundary_slot = head_state
            .current_epoch()
            .start_slot(T::EthSpec::slots_per_epoch());
        let epoch_boundary_root = if head_state.slot >= epoch_boundary_slot {
            head_root
        } else {
            *head_state.get_block_root(epoch_boundary_slot)?
        };
```

I think this is correct as is, `epoch_boundary_slot` is always equal to or less than `state.slot` and doing `state.get_block_root()` will correctly account for skip slots.

michaelsproul replied:
> epoch_boundary_slot is always equal to or less than state.slot and doing state.get_block_root() will correctly account for skip slots.

I agree with this part, but not the conclusion :sweat_smile: This observation implies `head_state.slot >= epoch_boundary_slot` will always evaluate to `true`, which means `epoch_boundary_root` will always be set to `head_root`, which is wrong.

Given that `head_state.slot < epoch_boundary_slot` is impossible, I think it should be an `==`, like:

```rust
        let epoch_boundary_slot = head_state
            .current_epoch()
            .start_slot(T::EthSpec::slots_per_epoch());
        let epoch_boundary_root = if head_state.slot == epoch_boundary_slot {
            head_root
        } else {
            *head_state.get_block_root(epoch_boundary_slot)?
        };
```

This might be a moot point anyway, if we switch to the last slot of the previous epoch, as per https://github.com/sigp/lighthouse/pull/1569/files#r494713541. Do you think that change is worth making?

paulhauner replied:
The changes in 5be8690 should make this irrelevant, thanks!

paulhauner commented on beacon_node/http_api/src/beacon_proposer_cache.rs:
```diff
@@ -0,0 +1,168 @@
+        // washed-out with old values.
+        if current_epoch != epoch {
+            return Err(warp_utils::reject::custom_bad_request(format!(
+                "requested epoch is {} but only current epoch {} is allowed",
```
So currently it's so that the value of the endpoint is easier to cache. We should really go back and add a slow, un-cached path for non-current requests, like we do for attester duties.

I've added this to #1434 so we can come back for it.

paulhauner commented on book/src/SUMMARY.md:
```diff
@@ -14,20 +14,15 @@
+	* [Validator Client API](./api-vc.md)
 * [Advanced Usage](./advanced.md)
     * [Database Configuration](./advanced_database.md)
+    * [Prometheus Metrics](./advanced_metrics.md)
+    * [Validator Inclusion APIs](./validator-inclusion.md)
```
I tried a new layout, let me know what you think :)

paulhauner commented on beacon_node/http_api/src/beacon_proposer_cache.rs:
```diff
@@ -0,0 +1,168 @@
+
+        let (head_root, head_block) =
+            Self::current_head_block(chain).map_err(warp_utils::reject::beacon_chain_error)?;
+        let epoch_boundary_root = head_block.target_root;
```
Great idea, I've implemented this here: 5be869095

paulhauner commented on beacon_node/http_api/src/beacon_proposer_cache.rs:
```diff
@@ -0,0 +1,168 @@
+        // We *must* skip forward to the current epoch to obtain valid proposer
+        // duties. We cannot skip to the previous epoch, like we do with
+        // attester duties.
+        while head_state.current_epoch() < current_epoch {
```
Oh yeah, good one. I've addressed this here: e4abccf5b

paulhauner commented on consensus/proto_array/src/proto_array_fork_choice.rs:
```diff
@@ -86,6 +88,12 @@ impl ProtoArrayForkChoice {
             // We are using the finalized_root as the target_root, since it always lies on an
             // epoch boundary.
             target_root: finalized_root,
+            // TODO: explain why this is safe.
```
Oh yeah, turns out it's probably not safe haha. I took the safer route here: e4070812a (will push shortly)

michaelsproul replied:
Looks good

paulhauner commented on consensus/serde_utils/src/u32_hex.rs:
```diff
@@ -0,0 +1,32 @@
+    let s: String = Deserialize::deserialize(deserializer)?;
+    let start = s
+        .as_str()
+        .get(2..)
```
I've cherry-picked that commit, thanks!

paulhauner commented on beacon_node/beacon_chain/src/naive_aggregation_pool.rs:
```diff
@@ -116,7 +121,12 @@ impl<E: EthSpec> AggregatedAttestationMap<E> {
+    /// Returns an aggregated `Attestation` with the given `root`, if any.
+    pub fn get_by_root(&self, root: &AttestationDataRoot) -> Option<&Attestation<E>> {
+        self.map.get(root)
     }
```
Turns out the `get` result was completely redundant! Thanks

paulhauner commented:
Alrighty, I believe I have addressed all PR comments! There's quite a few, so please let me know if I've missed any or perhaps glossed over anything important.

Thank to @michaelsproul and @realbigsean for your reviews! Both reviews made significant improvements :) I believe this diff contains the changes between when you reviewed and now: https://github.com/sigp/lighthouse/compare/006ca02a5fa0c1103e4a0f793e89763352265a33..31b0ac1e50998249f3fe0798e3f19283c8cae792

I'll leave CI to run and f it fails I'll follow it up.

michaelsproul reviewed (approved):
Happy to merge to v0.3.0 once conflicts are resolved! :tada:

paulhauner commented:
Yay! Thanks all involved! I'll try an optimistic bors and see how it goes!

bors r+
